syntax = "proto3";

import "google/protobuf/wrappers.proto";

package proto.client;

// Message sent from client applications to the Universal slaves.
message ClientRequest {
  enum Type {
    READ = 0;
    WRITE = 1;
  }

  string request_id = 1;
  Type request_type = 2;
  google.protobuf.StringValue key = 3;
  google.protobuf.StringValue value = 4;
  google.protobuf.UInt64Value timestamp = 5;
}

// Response sent to a client from the Universal slaves.
message ClientResponse {
  enum Code {
    SUCCESS = 0;
    ERROR = 1;
  }

  google.protobuf.StringValue value = 1;
  Code error_code = 2;
}

message ClientMessage {
  oneof client_message {
    ClientRequest request = 1;
    ClientResponse response = 2;
  }
}
